src/common/async/yield_context.h

```diff
@@ -27,27 +34,180 @@
 /// will, when passed a non-empty yield context, suspend this coroutine instead
 /// of the blocking the thread of execution
 class optional_yield {
-  boost::asio::io_context *c = nullptr;
-  spawn::yield_context *y = nullptr;
+  static std::mutex lock;
+  static std::function<void()> checker;
+
+  using cy = std::pair<boost::asio::io_context*, spawn::yield_context>;
+
+  std::variant<cy, ceph::async::use_blocked_t> v;
```

Comment:
my main concern here is that it's now storing `spawn::yield_context` by value instead of pointer. it contains a `std::weak_ptr`, so copies will involve atomic ref counting. in radosgw, we pass `optional_yield` by value all over the place, so this is a big deal

it looks like the requirement to store by value comes from `operator[]`, which needs to construct a new one and assign it in place. maybe we could get around this by having `operator[]` return some other type? but that could mean duplicating all of the async_result machinery for another type. what do you think?

Reply:
We could try a tripartite variant and use a pointer unless we have to use a value.

Given that most of the functions we use this on do their own fiddling with ec, that might not be an issue, since if we did bind one to it, we'd just ignore it. So making that usable only when we're passing to something expecting a CompletionToken doesn't seem bad.

I think I can avoid the duplication being too bad.

So having [] return a different type sounds good overall.

Reply:
> We could try a tripartite variant and use a pointer unless we have to use a value.

that sounds like the easiest solution